File: ATV-4/arvore.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node* left;
    struct Node* right;
} Node;

// Função para criar um novo nó
Node* createNode(int data) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->data = data;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}

// Função para inserir um nó na árvore
Node* insertNode(Node* root, int data) {
    if (root == NULL) {
        root = createNode(data);
    } else if (data <= root->data) {
        root->left = insertNode(root->left, data);
    } else {
        root->right = insertNode(root->right, data);
    }
    return root;
}

// Função para imprimir a árvore em ordem
void printInOrder(Node* root) {
    if (root == NULL) {
        return;
    }
    printInOrder(root->left);
    printf("%d ", root->data);
    printInOrder(root->right);
}

// Função para imprimir a árvore em pré-ordem
void printPreOrder(Node* root) {
    if (root == NULL) {
        return;
    }
    printf("%d ", root->data);
    printPreOrder(root->left);
    printPreOrder(root->right);
}

// Função para imprimir a árvore em pós-ordem
void printPostOrder(Node* root) {
    if (root == NULL) {
        return;
    }
    printPostOrder(root->left);
    printPostOrder(root->right);
    printf("%d ", root->data);
}

// Função para encontrar o nó mínimo na subárvore direita
Node* findMinNode(Node* node) {
    while (node->left != NULL) {
        node = node->left;
    }
    return node;
}

// Função para remover um nó da árvore
Node* removeNode(Node* root, int data) {
    if (root == NULL) {
        return root;
    } else if (data < root->data) {
        root->left = removeNode(root->left, data);
    } else if (data > root->data) {
        root->right = removeNode(root->right, data);
    } else {
        // Caso 1: Nó sem filhos
        if (root->left == NULL && root->right == NULL) {
            free(root);
            root = NULL;
        }
        // Caso 2: Nó com um filho
        else if (root->left == NULL) {
            Node* temp = root;
            root = root->right;
            free(temp);
        } else if (root->right == NULL) {
            Node* temp = root;
            root = root->left;
            free(temp);
        }
        // Caso 3: Nó com dois filhos
        else {
            Node* temp = findMinNode(root->right);
            root->data = temp->data;
            root->right = removeNode(root->right, temp->data);
        }
    }
    return root;
}

// Função para liberar a memória ocupada pela árvore
void clearTree(Node* root) {
    if (root == NULL) {
        return;
    }
    clearTree(root->left);
    clearTree(root->right);
    free(root);
}

int main() {
    Node* root = NULL;
    int choice, data;

    while (1) {
        printf("\n");
        printf("Selecione uma opcao:\n");
        printf("1. Inserir um no na arvore\n");
        printf("2. Listar a arvore (ordem, pre e pos)\n");
        printf("3. Remover um no da arvore\n");
        printf("4. Esvaziar a arvore\n");
        printf("5. Sair\n");
        printf("Opcao: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                printf("Digite o valor a ser inserido: ");
                scanf("%d", &data);
                root = insertNode(root, data);
                break;
            case 2:
                printf("Arvore em ordem: ");
                printInOrder(root);
                printf("\n");
                printf("Arvore em pre-ordem: ");
                printPreOrder(root);
                printf("\n");
                printf("Arvore em pos-ordem: ");
                printPostOrder(root);
                printf("\n");
                break;
            case 3:
                printf("Digite o valor a ser removido: ");
                scanf("%d", &data);
                root = removeNode(root, data);
                break;
            case 4:
                clearTree(root);
                root = NULL;
                printf("A arvore foi esvaziada.\n");
                break;
            case 5:
                clearTree(root);
                printf("Encerrando o programa.\n");
                exit(0);
            default:
                printf("Opcao invalida! Tente novamente.\n");
        }
    }

    return 0;
}

File: ATV-4/arvore.py
class No:
    def __init__(self, valor):
        self.valor = valor
        self.esquerda = None
        self.direita = None

class Arvore:
    def __init__(self):
        self.raiz = None

    def inserir(self, valor):
        if self.raiz is None:
            self.raiz = No(valor)
        else:
            self._inserir_recursivo(valor, self.raiz)

    def _inserir_recursivo(self, valor, no_atual):
        if valor < no_atual.valor:
            if no_atual.esquerda is None:
                no_atual.esquerda = No(valor)
            else:
                self._inserir_recursivo(valor, no_atual.esquerda)
        else:
            if no_atual.direita is None:
                no_atual.direita = No(valor)
            else:
                self._inserir_recursivo(valor, no_atual.direita)

    def listar_ordem(self):
        self._listar_ordem_recursivo(self.raiz)

    def _listar_ordem_recursivo(self, no_atual):
        if no_atual:
            self._listar_ordem_recursivo(no_atual.esquerda)
            print(no_atual.valor, end=" ")
            self._listar_ordem_recursivo(no_atual.direita)

    def listar_preordem(self):
        self._listar_preordem_recursivo(self.raiz)

    def _listar_preordem_recursivo(self, no_atual):
        if no_atual:
            print(no_atual.valor, end=" ")
            self._listar_preordem_recursivo(no_atual.esquerda)
            self._listar_preordem_recursivo(no_atual.direita)

    def listar_posordem(self):
        self._listar_posordem_recursivo(self.raiz)

    def _listar_posordem_recursivo(self, no_atual):
        if no_atual:
            self._listar_posordem_recursivo(no_atual.esquerda)
            self._listar_posordem_recursivo(no_atual.direita)
            print(no_atual.valor, end=" ")

    def remover(self, valor):
        self.raiz = self._remover_recursivo(valor, self.raiz)

    def _remover_recursivo(self, valor, no_atual):
        if no_atual is None:
            return no_atual
        elif valor < no_atual.valor:
            no_atual.esquerda = self._remover_recursivo(valor, no_atual.esquerda)
        elif valor > no_atual.valor:
            no_atual.direita = self._remover_recursivo(valor, no_atual.direita)
        else:
            if no_atual.esquerda is None:
                return no_atual.direita
            elif no_atual.direita is None:
                return no_atual.esquerda
            else:
                sucessor = self._encontrar_sucessor(no_atual.direita)
                no_atual.valor = sucessor.valor
                no_atual.direita = self._remover_recursivo(sucessor.valor, no_atual.direita)
        return no_atual

    def _encontrar_sucessor(self, no_atual):
        while no_atual.esquerda:
            no_atual = no_atual.esquerda
        return no_atual

    def esvaziar(self):
        self.raiz = None


def exibir_menu():
    print("Escolha uma opção:")
    print("1 - Inserir um nó")
    print("2 - Listar a árvore (em ordem)")
    print("3 - Listar a árvore (pré-ordem)")
    print("4 - Listar a árvore (pós-ordem)")
    print("5 - Remover um nó")
    print("6 - Esvaziar a árvore")
    print("7 - Sair")
    opcao = int(input("Opção: "))
    return opcao


arvore = Arvore()

while True:
    opcao = exibir_menu()

    if opcao == 1:
        valor = int(input("Digite o valor a ser inserido: "))
        arvore.inserir(valor)
        print("Valor inserido com sucesso!")
    elif opcao == 2:
        print("Árvore (em ordem):")
        arvore.listar_ordem()
        print()
    elif opcao == 3:
        print("Árvore (pré-ordem):")
        arvore.listar_preordem()
        print()
    elif opcao == 4:
        print("Árvore (pós-ordem):")
        arvore.listar_posordem()
        print()
    elif opcao == 5:
        valor = int(input("Digite o valor a ser removido: "))
        arvore.remover(valor)
        print("Valor removido com sucesso!")
    elif opcao == 6:
        arvore.esvaziar()
        print("Árvore esvaziada!")
    elif opcao == 7:
        print("Saindo...")
        break
    else:
        print("Opção inválida. Tente novamente!")
